RecoTracker/FinalTrackSelectors/plugins/TrackCollectionMerger.cc

```diff
@@ -0,0 +1,347 @@
+      m_allowFirstHitShare(conf.getParameter<bool>("allowFirstHitShare"))
+{
+      for (auto const & it : conf.getParameter<std::vector<edm::InputTag> >("trackProducers") )
+	srcColls.emplace_back(it,consumesCollector());
+      for (auto const & it : conf.getParameter<std::vector<std::string> >("inputClassifiers")) {
+	srcMVAs.push_back(consumes<MVACollection>(edm::InputTag(it,"MVAValues")));
+	srcQuals.push_back(consumes<QualityMaskCollection>(edm::InputTag(it,"QualityMasks")));
+      }
+
+      assert(srcColls.size()==srcQuals.size());
+      
+
+      produces<MVACollection>("MVAValues");
+      produces<QualityMaskCollection>("QualityMasks");
+
+    }
+      
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+      edm::ParameterSetDescription desc;
+      desc.add<std::vector<edm::InputTag> >("trackProducers",std::vector<edm::InputTag>());
+      desc.add<std::vector<std::string> >("inputClassifiers",std::vector<std::string>());
+      desc.add<double>("shareFrac",.19);
+      desc.add<double>("foundHitBonus",10.);
+      desc.add<double>("lostHitPenalty",5.);
+      desc.add<unsigned int>("minShareHits",2);
+      desc.add<bool>("allowFirstHitShare",true);
+      desc.add<std::string>("minQuality","loose");
+      descriptions.add("TrackCollectionMerger", desc);
+    }
+
+   private:
+
+
+    TrackCollectionCloner collectionCloner;
+    std::vector<TrackCollectionCloner::Tokens> srcColls;
+
+    using MVACollection = std::vector<float>;
+    using QualityMaskCollection = std::vector<unsigned char>;
+    
+    using IHit = std::pair<unsigned int, TrackingRecHit const *>;
+    using IHitV = std::vector<IHit>;
+    
+    std::vector<edm::EDGetTokenT<MVACollection>> srcMVAs;
+    std::vector<edm::EDGetTokenT<QualityMaskCollection>> srcQuals;
+
+	
+    float m_foundHitBonus;
+    float m_lostHitPenalty;
+    float m_shareFrac;
+    unsigned int m_minShareHits;
+    reco::TrackBase::TrackQuality m_minQuality;
+    bool  m_allowFirstHitShare;
+    
+    virtual void produce(edm::StreamID, edm::Event& evt, const edm::EventSetup&) const override;
+    
+
+    bool areDuplicate(IHitV const& rh1, IHitV const& rh2) const;
+      
+
+  };
+
+}
+  
+#include "CommonTools/Utils/interface/DynArray.h"
+
+
+namespace {
+  
+  void  TrackCollectionMerger::produce(edm::StreamID, edm::Event& evt, const edm::EventSetup&) const {
+
+    TrackCollectionCloner::Producer producer(evt, collectionCloner);
+
+    // load collections
+    auto collsSize = srcColls.size();
+    auto rSize=0U;
+    declareDynArray(reco::TrackCollection const *, collsSize, trackColls);
+    for (auto i=0U; i< collsSize; ++i) {
+      trackColls[i] = &srcColls[i].tracks(evt);
+      rSize += (*trackColls[i]).size();
+    }
+
+
+    unsigned char qualMask = ~0;
+    if (m_minQuality!=reco::TrackBase::undefQuality) qualMask = 1<<m_minQuality; 
+    
+    
+    // load tracks
+    initDynArray(unsigned int,collsSize,nGoods,0);
```

Comment:
what a pain.
In any case it is true that it can happen (in case of upstream errors)
The code is complex enough that adding a  if (srcColls.empty()) bailout will not harm...